design-docs/Defer draft API.md

```diff
@@ -0,0 +1,852 @@
+will break such implementations.
+
+### HttpEngine / HttpInterceptor / HttpResponse
+
+Here’s the current definitions of these 3 classes:
+
+```kotlin
+interface HttpEngine {
+  suspend fun execute(request: HttpRequest): HttpResponse
+  fun dispose()
+}
+
+interface HttpInterceptor {
+  suspend fun intercept(request: HttpRequest, chain: HttpInterceptorChain): HttpResponse
+  fun dispose() {}
+}
+
+class HttpResponse {
+  val statusCode: Int,
+  val headers: List<HttpHeader>,
+  val body: BufferedSource?,
+}
+```
+
+As we can see the contracts are designed around 1 full response being returned from a request, whereas when
+using `@defer` this is no longer appropriate: several chunks can be returned.
+
+To address this let’s augment `HttpResponse` with a Flow representing the chunks:
+
+```kotlin
+class HttpResponse {
+  val statusCode: Int,
+  val headers: List<HttpHeader>,
+  val body: BufferedSource?,
+  val chunks: Flow<HttpChunk>, // <- New!
+}
+
+class HttpChunk {
+  val headers: List<HttpHeader>,
+  val body: BufferedSource?,
+}
+```
+
+This introduces `HttpChunk` which is a light version of `HttpResponse`.
+
+Notes:
+
+- in the non-chunked (`@defer` not used) case, `chunks` will be an empty Flow
+- when `chunks` is not empty, the response `body` is `null`
+- any `HttpInterceptor` implementation that manipulates the response body won’t automatically work on chunks. Not sure
+  if this is a common use-case but it will break / not work as expected.
+
+### Http Cache
+
+The http cache can still be used in `@defer` scenarios.
+
+A response will be cached after all its chunks have been received.
+
+When a response comes from the cache, the chunks will be available immediately.
+
+### Batching
+
+At this point it is unclear whether batching in a `@defer` scenario is useful. For now, we can decide that batching is
+automatically opted-out for any operations using `@defer`.
+
+### Normalized Cache
+
+Several options are possible:
+
+**A. Incremental data is cached as it is being received**
+
+Pros:
+
+- When using `watch` the UI can be as reactive as possible and reflect the received data in real time
+
+Cons:
+
+- When using `query` with `FetchPolicy.CacheOnly` partial data can be returned, with no way to know this
+- A potentially undesirable experience with `watch` when calls are done in parallel. For instance a collector could receive in this order:
+    1. A Computer *(from request **a**)*
+    2. A Computer, with computerFields *(from request **a**)*
+    3. A Computer *(from request **b**)*
+    4. A Computer, with computerFields, with screenFields *(from request **a**)*
+    5. A Computer, with computerFields *(from request **b**)*
+    6. A Computer, with computerFields, with screenFields *(from request **b**)*
+
+Open questions
+
+- What happens if the key fields are deferred? Maybe this could be forbidden via a compile time check?
```

Comment:
Yikes 😱 . The compile time check should even be more complicated than this...

This is allowed:

```graphql
query Query {
  ... on Computer @defer {
    computer {
      id
      cpu
      year
    }
  }
}
```

This is not:


```graphql
query Query {
  ... on Computer @defer {
    computer {
      id
    }
  }
  computer {
     # we get cpu and year without the id yet so we cannot write it
      cpu
      year
  }
}
```

Or maybe this is allowed but the write only happens once the id is received 🤔

Reply:
> Or maybe this is allowed but the write only happens once the id is received 🤔

Oh yeah If doable, this sounds like the best behavior for users (it "just works")!